ios/Classes/TiAirshipModule.m

```diff
@@ -165,19 +187,33 @@ - (void)associateIdentifier:(id)args {
     [[UAirship shared].analytics associateDeviceIdentifiers:identifiers];
 }
 
-- (void)addCustomEvent:(id)eventPayload {
-    NSString *customEventString = [TiUtils stringValue:[eventPayload objectAtIndex:0]];
-    UA_LDEBUG(@"Add custom event: %@", customEventString);
+- (void)addCustomEvent:(id)args {
+    ENSURE_ARG_COUNT(args, 1);
+    NSDictionary *payload;
 
-    if (customEventString.length == 0) {
-        UA_LERR(@"Missing event payload.");
-        return;
+    if ([args[0] isKindOfClass:[NSString class]]) {
+        NSString *customEventString = [TiUtils stringValue:[args objectAtIndex:0]];
+        if (customEventString.length == 0) {
+            UA_LERR(@"Missing event payload.");
+            return;
+        }
+        NSError *jsonError;
+        NSData *data = [customEventString dataUsingEncoding:NSUTF8StringEncoding];
+        payload = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&jsonError];
+        if (jsonError) {
+            UA_LERR(@"Invalid custom event payload %@", customEventString);
+            return;
+        }
+    } else {
+        ENSURE_SINGLE_ARG(args, NSDictionary);
```

Comment:
I'm a little confused by this, should it be `ENSURE_SINGLE_ARG(args[0], NSDictionary)`?  How can args have an index and be a dictionary at the same time?

Reply:
Magic.